cucumber-picocontainer/src/main/java/io/cucumber/picocontainer/PicoFactory.java

```diff
@@ -15,36 +16,45 @@
 public final class PicoFactory implements ObjectFactory {
 
     private final Set<Class<?>> classes = new HashSet<>();
-    private final MutablePicoContainer pico = new PicoBuilder()
-            .withCaching()
-            .withLifecycle()
-            .build();
-
-    public PicoFactory() {
-        this.pico.start();
-    }
+    private MutablePicoContainer pico;
 
     private static boolean isInstantiable(Class<?> clazz) {
         boolean isNonStaticInnerClass = !Modifier.isStatic(clazz.getModifiers()) && clazz.getEnclosingClass() != null;
         return Modifier.isPublic(clazz.getModifiers()) && !Modifier.isAbstract(clazz.getModifiers())
                 && !isNonStaticInnerClass;
     }
 
+    @Override
     public void start() {
-        // do nothing (was already started in constructor)
+        if (pico == null) {
+            pico = new PicoBuilder()
+                    .withCaching()
+                    .withLifecycle()
+                    .build();
+            for (Class<?> clazz : classes) {
+                pico.addComponent(clazz);
+            }
+        } else {
+            // we already get a pico container which is in "disposed" lifecycle,
+            // so recycle it by defining a new lifecycle and removing all
+            // instances
+            pico.setLifecycleState(new DefaultLifecycleState());
+            pico.getComponentAdapters()
+                    .forEach(cached -> ((Cached<?>) cached).flush());
+        }
+        pico.start();
```

Comment:
Performance wise, how do the different branches compare? If there still an appreciable performance gain?

Reply:
Yes, performance is still good. JMH micro-benchmark shows:

|Benchmark |                     Mode | Cnt |      Score |      Error | Units|
|--------------|--------------------|-----|------------|---------|-----------|
|PicoFactoryBenchmark.v7_11_1 | thrpt |  25  | 24015,595 |±  1681,477  |ops/s|
|PicoFactoryBenchmark.v7_12_1|  thrpt |  25 | 769958,897 |± 65377,491 | ops/s|
|PicoFactoryBenchmark.v7_13 (this PR)    |thrpt  | 25 | 709328,227 |± 12756,281  |ops/s|

The benchmark consists in instanciating a new PicoFactory ouside the benchmark, then adding 7 classes (containing 183 step defs in total), then start the factory, get one instance and stop the factory:

    public class PicoFactoryBenchmark {
      @Benchmark
      public List<Object> v7_11_1(PicoFactoryPlan plan) {
          return benchmark(plan.picoFactory7_11_1);
      }
  
      @Benchmark
      public List<Object> v7_12_1(PicoFactoryPlan plan) {
          return benchmark(plan.picoFactory7_12_1);
      }
      @Benchmark
      public List<Object> v7_13(PicoFactoryPlan plan) {
          return benchmark(plan.picoFactory7_13);
      }
      private static List<Object> benchmark(ObjectFactory objectFactory) {
          List<Object> instances = new ArrayList<>();
          objectFactory.addClass(A.class);
          objectFactory.addClass(B.class);
          objectFactory.addClass(C.class);
          objectFactory.addClass(D.class);
          objectFactory.addClass(F.class);
          objectFactory.addClass(F.class);
          objectFactory.addClass(G.class);
  
          objectFactory.start();
          instances.add(objectFactory.getInstance(A.class));
          objectFactory.stop();
  
          return instances;
      }
  
    }

I think the version from Cucumber 7.12.1 is about 10% faster than the new code for 7.13 because the pico-container is started in the constructor (that is, outside the benchmark). But this code for 7.13 is still 30 times faster than in Cucumber 7.11.1.